### PR TITLE
Disallow writes on materialized views

### PR DESCRIPTION
**Pull request by findinpath**
## Description

Disallow `UPDATE` , `MERGE` , `DELETE` on materialized views.

`INSERT` and `TRUNCATE` have been handled already.

## Non-technical explanation

Disallow `UPDATE` , `MERGE` , `DELETE` on materialized views.

## Release notes

(x) This is not user-visible and no release notes are required.
( ) Release notes are required, please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

File: core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java [hunk 1]
```diff
@@ -748,6 +748,9 @@ protected Scope visitDelete(Delete node, Optional<Scope> scope)
         {
             Table table = node.getTable();
             QualifiedObjectName originalName = createQualifiedObjectName(session, table, table.getName());
+            if (metadata.isMaterializedView(session, originalName)) {
+                throw semanticException(NOT_SUPPORTED, node, "Deleting from materialized views is not supported");
```
[the remark this answers is not in the document]
Comment (findinpath) [reply core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java:752]: The statement was failing with "Table '%s' does not exist" exception

File: core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java [hunk 2]
```diff
@@ -2690,6 +2693,9 @@ protected Scope visitUpdate(Update update, Optional<Scope> scope)
         {
             Table table = update.getTable();
             QualifiedObjectName originalName = createQualifiedObjectName(session, table, table.getName());
+            if (metadata.isMaterializedView(session, originalName)) {
+                throw semanticException(NOT_SUPPORTED, update, "Updating through materialized views is not supported");
```
Comment (phd3) [inline core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java:2697]: nit: `updating through` -> `updating` would be more apprioriate

File: core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java [hunk 3]
```diff
@@ -2793,6 +2799,9 @@ protected Scope visitMerge(Merge merge, Optional<Scope> scope)
             Relation relation = merge.getTarget();
             Table table = getMergeTargetTable(relation);
             QualifiedObjectName tableName = createQualifiedObjectName(session, table, table.getName());
+            if (metadata.isMaterializedView(session, tableName)) {
+                throw semanticException(NOT_SUPPORTED, merge, "MERGE INTO a materialized view is not supported");
```
Comment (phd3) [inline core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java:2803]: nit: INTO here reads as SQL syntax clause and preposition both : ) how about 

MERGE INTO is not supported for a materialized view. Or 
merging into a materialized view is not supported.

File: core/trino-main/src/test/java/io/trino/sql/analyzer/TestAnalyzer.java [hunk 4]
```diff
@@ -3890,6 +3890,40 @@ public void testInvalidDelete()
...
+    @Test
+    public void testInvalidUpdate()
+    {
+        assertFails("UPDATE foo SET bar = 'test'")
+                .hasErrorCode(TABLE_NOT_FOUND)
+                .hasMessage("line 1:1: Table 'tpch.s1.foo' does not exist");
+        assertFails("UPDATE v1 SET a = 2")
+                .hasErrorCode(NOT_SUPPORTED)
+                .hasMessage("line 1:1: Updating through views is not supported");
+        assertFails("UPDATE v1 SET a = 2 WHERE a = 1")
```
Comment (phd3) [inline core/trino-main/src/test/java/io/trino/sql/analyzer/TestAnalyzer.java:3907]: is there a purpose to have both test `UPDATE v1 SET a = 2` and `UPDATE v1 SET a = 2 WHERE a = 1`?

**Review by phd3 — COMMENTED** [review]
thanks for the fix @findinpath, this looks good to me.. A couple very minor comments

File: core/trino-main/src/test/java/io/trino/sql/analyzer/TestAnalyzer.java (hunk 4 above)
Comment (findinpath) [reply core/trino-main/src/test/java/io/trino/sql/analyzer/TestAnalyzer.java:3907]: No, actually not. I was wondering as well while I adapted the method from `testInvalidUpdate()`.

**Comment by findepi** [comment]
@findinpath mind the CI

**Comment by findinpath** [comment]
The CI failure is related to a flaky test and documented into https://github.com/trinodb/trino/issues/13889

**Review by findepi — APPROVED** [review]

**Pull request merged**